### PR TITLE
[notifications] Fix older notifications be overwritten issue

### LINKED ISSUE #17966: [expo-notifications] SDK 45 addNotificationResponseReceivedListener always return the latest notification received
### Summary

On bare workflow and **ONLY on Android 12** when you tap a notification `Notifications.addNotificationResponseReceivedListener` always return the latest notification received, not the notification tapped.

I tested this on Android 8.1 and works as expected, seems that it's a bug only on **Android 12**.

I created a demo using `scheduleNotificationAsync`, but the same happens when you send notifications throw FCM.

```
expo: 45.0.6
expo-notifications: 0.15.3
```

**Demo:**
ezgif com-gif-maker(2)

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

bare

### What platform(s) does this occur on?

Android

### SDK Version (managed workflow only)

45

### Environment

```
  expo-env-info 1.0.3 environment info:
    System:
      OS: macOS 12.4
      Shell: 5.8.1 - /bin/zsh
    Binaries:
      Node: 16.15.1 - /usr/local/bin/node
      npm: 8.11.0 - /usr/local/bin/npm
      Watchman: 4.9.0 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.11.3 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 21.4, iOS 15.5, macOS 12.3, tvOS 15.4, watchOS 8.5
    IDEs:
      Android Studio: 2021.2 AI-212.5712.43.2112.8609683
      Xcode: 13.4.1/13F100 - /usr/bin/xcodebuild
    npmPackages:
      expo: ~45.0.0 => 45.0.6
      react: 17.0.2 => 17.0.2
      react-dom: 17.0.2 => 17.0.2
      react-native: 0.68.2 => 0.68.2
      react-native-web: 0.17.7 => 0.17.7
    npmGlobalPackages:
      eas-cli: 0.22.0
      expo-cli: 5.4.11
    Expo Workflow: bare
```

### Reproducible demo

It's super easy to reproduce, just make sure that you test it on a device with **Android 12** (I'm using a Google Pixed 4a)

1. `expo init expo-push-issue`
2. select `Bare workflow -> minimal`
3. `cd expo-push-issue`
4. `expo install expo-notifications`
5. Copy code of App.js from here https://github.com/RodolfoGS/expo-push-issue/blob/main/App.js
6. `npm run android`
7. Tap **press to send notification** button 3 times
8. Open the notification with the title **Test notification number 1**
9. You'll see an alert with the title of the last notification sent

If you prefer I created a repo with the previous steps: https://github.com/RodolfoGS/expo-push-issue
1. `git clone git@github.com:RodolfoGS/expo-push-issue.git`
2. `cd expo-push-issue`
3. `npm run android`
4. Tap **press to send notification** button 3 times
5. Open the notification with the title **Test notification number 1**
6. You'll see an alert with the title of the last notification sent

### PR DESCRIPTION
**Pull request by Kudo**
# Why

when there are multiple notifications, the latest one will overwrite previous' intent extra data. `useLastNotificationResponse` returns the latest received notification but not the clicked notification.

fixed https://github.com/expo/expo/issues/17966 https://github.com/expo/expo/issues/17531#issuecomment-1160495971

# How

- the previous PendingIntent be overwritten because `PendingIntent.FLAG_UPDATE_CURRENT`. we also used `PendingIntent.FLAG_UPDATE_CURRENT` for previous implementation, the main difference is that for `PendingIntent.getActivities`, it is slightly different to check whether two PendingIntent are equal. i noticed that OneSignal reverted their implementation of PendingIntent.getActivities aka the Reverse Activity Trampoline because the problem for Xiaomi devices. in this pr, i changed the implementation to

```
notification pendingIntent -> hidden background Activity (NotificationForwarderActivity) -> foreground Activity -> broadcast receiveResponse -> finishe NotificationForwarderActivity
``` 

- remove `foregroundActivityIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_MULTIPLE_TASK` that cause #17960

- backport changes to versioned code

# Test Plan

follow #17966 to test multiple notifications and the `Notifications.addNotificationResponseReceivedListener` returned content

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/NotificationForwarderActivity.kt [hunk 1]
```diff
@@ -15,6 +16,14 @@ class NotificationForwarderActivity : Activity() {
     super.onCreate(savedInstanceState)
     val broadcastIntent =
       NotificationsService.createNotificationResponseBroadcastIntent(applicationContext, intent.extras)
+
+    val foregroundLaunchActivityIntent = ExpoHandlingDelegate.getForegroundLaunchActivityIntent(this)
+    if (foregroundLaunchActivityIntent != null) {
+      val notificationResponse = NotificationsService.getNotificationResponseFromBroadcastIntent(broadcastIntent)
+      NotificationsService.setNotificationResponseToIntent(foregroundLaunchActivityIntent, notificationResponse)
+      startActivity(foregroundLaunchActivityIntent)
```
Comment (bbarthec) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/NotificationForwarderActivity.kt:24]: I see this piece of code is also present in `ExpoHandlingDelegate` itself: https://github.com/expo/expo/blob/660574385842ff87d0811be4b9968b8693317038/packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/ExpoHandlingDelegate.kt#L133-L138

Maybe we should unify these paths? From what I can see `NotificationForwarderActivity` is only a proxy for broadcasting/delegating heavy stuff to some other classes, so I guess it should not duplicate the code from `ExpoHandlingDelegate`

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/ExpoHandlingDelegate.kt [hunk 2]
```diff
@@ -67,15 +67,22 @@ class ExpoHandlingDelegate(protected val context: Context) : HandlingDelegate {
...
+      return PendingIntent.getActivity(context, requestCode, backgroundActivityIntent, intentFlags)
+    }
+
+    fun getForegroundLaunchActivityIntent(context: Context): Intent? {
+      val intent = getNotificationActionLauncher(context)
+        ?: ExpoHandlingDelegate.getMainActivityLauncher(context) ?: run {
+        Log.w("expo-notifications", "No launch intent found for application. Interacting with the notification won't open the app. The implementation uses `getLaunchIntentForPackage` to find appropriate activity.")
+        return null
+      }
+      return intent
```
Comment (bbarthec) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/ExpoHandlingDelegate.kt:85]: Let's have each elvis operator start a new line 😉 
```suggestion
    fun getForegroundLaunchActivityIntent(context: Context): Intent? =
      getNotificationActionLauncher(context)
        ?: ExpoHandlingDelegate.getMainActivityLauncher(context)
        ?: run {
          Log.w("expo-notifications", "No launch intent found for application. Interacting with the notification won't open the app. The implementation uses `getLaunchIntentForPackage` to find appropriate activity.")
          return null
        }
```

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/NotificationForwarderActivity.kt [hunk 3]
```diff
@@ -15,6 +16,14 @@ class NotificationForwarderActivity : Activity() {
     super.onCreate(savedInstanceState)
     val broadcastIntent =
       NotificationsService.createNotificationResponseBroadcastIntent(applicationContext, intent.extras)
+
+    val foregroundLaunchActivityIntent = ExpoHandlingDelegate.getForegroundLaunchActivityIntent(this)
+    if (foregroundLaunchActivityIntent != null) {
```
Comment (bbarthec) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/NotificationForwarderActivity.kt:21]: Up to you 😉 
```suggestion
    ExpoHandlingDelegate.getForegroundLaunchActivityIntent(this)?.let { foregroundLaunchActivityIntent ->
```

**Review by bbarthec — COMMENTED** [review]
Looking good to me 👍 
Left one suggestion and one question 👀

**Review by bbarthec — APPROVED** [review]
Superb! Thank you for taking care of it! 🚀 🤩

**Comment by code4muktesh** [comment]
Not working when app in background.
On Android real device _useLastNotificationResponse_ returns null after clicking on a notification

**Pull request merged**